File: examples/ts/Scripts/ChatScene.ts
/// <reference path="../fairygui.d.ts" />

import { Callable, RegEx } from "godot";
import { DemoSceneBase } from "./DemoSceneBase";

interface ChatMessage {
    sender: string;
    sender_icon: string;
    text: string;
    from_me: boolean;
}

export default class ChatScene extends DemoSceneBase {
    private _list: GList | null = null;
    private _inputField: GTextField | null = null;
    private _messages: ChatMessage[] = [];
    private _emojiSelectUi: GComponent | null = null;

    protected ContinueInit(): void {
        UIPackage.addPackage("res://Resources/UI/Emoji");

        this._view = UIPackage.createObject("Emoji", "Main");
        this._groot!.addChild(this._view);

        this._list = this._view.getChild("list") as GList | null;
        if (this._list != null) {
            this._list.setVirtual();
            this._list.setItemRenderer(Callable.create(this._renderListItem.bind(this)));
            this._list.setItemProvider(Callable.create(this._getListItemResource.bind(this)));
        }

        this._inputField = this._view.getChild("input") as GTextField | null;
        if (this._inputField != null) {
            this._inputField.addEventListener(
                UIEventDispatcher.SUBMIT,
                Callable.create(this._onSubmit.bind(this)),
            );
        }

        const sendBtn = this._view.getChild("btnSend");
        if (sendBtn != null) {
            sendBtn.addClickListener(Callable.create(this._onClickSendBtn.bind(this)));
        }

        const emojiBtn = this._view.getChild("btnEmoji");
        if (emojiBtn != null) {
            emojiBtn.addClickListener(Callable.create(this._onClickEmojiBtn.bind(this)));
        }

        this._emojiSelectUi = UIPackage.createObject("Emoji", "EmojiSelectUI");
        if (this._emojiSelectUi != null) {
            const emojiList = this._emojiSelectUi.getChild("list");
            if (emojiList != null) {
                emojiList.addEventListener(
                    UIEventDispatcher.CLICKITEM,
                    Callable.create(this._onClickEmoji.bind(this)),
                );
            }
        }

        this._addMsg("Unity", "r0", "Hello!", true);
    }

    private _onClickSendBtn(): void {
        if (this._inputField == null) {
            return;
        }
        const msg = this._inputField.getText();
        if (msg.length === 0) {
            return;
        }
        this._addMsg("Unity", "r0", msg, true);
        this._inputField.setText("");
    }

    private _onClickEmojiBtn(): void {
        if (this._emojiSelectUi == null) {
            return;
        }
        this._groot!.showPopupSimple(this._emojiSelectUi);
    }

    private _onClickEmoji(ctx: FGUIEventContext | null = null): void {
        let item: GuiObject | null = null;
        if (ctx != null) {
            item = ctx.getData() as GuiObject | null;
        }
        if (item == null) {
            item = this._groot!.getTouchTarget();
        }
        if (item == null || this._inputField == null) {
            return;
        }
        const tag = item.getText();
        if (tag.length === 0) {
            return;
        }
        this._inputField.setText(`${this._inputField.getText()}[:${tag}]`);
    }

    private _onSubmit(): void {
        this._onClickSendBtn();
    }

    private _parseEmoji(text: string): string {
        const regex = new RegEx();
        regex.compile("\\[:\\s*(\\w+)\\]");
        while (true) {
            const result = regex.search(text);
            if (result == null) {
                break;
            }
            const tag = result.getString(1);
            text = text.replace(result.getString(0), `<img src='ui://Emoji/${tag}'/>`);
        }
        return text;
    }

    private _renderListItem(index: number, obj: GComponent): void {
        if (index >= this._messages.length) {
            return;
        }
        const msg = this._messages[index];
        obj.icon = `ui://Emoji/${msg.sender_icon}`;

        const tf = obj.getChild("msg");
        if (tf instanceof GRichTextField) {
            tf.setUBBEnabled(false);
            tf.setWidth(tf.getInitSize().x);
            tf.setText(this._parseEmoji(msg.text));
            tf.setWidth(tf.getTextSize().x);
        }

        const nameObj = obj.getChild("name");
        if (nameObj != null && !msg.from_me) {
            nameObj.setText(msg.sender);
        }
    }

    private _getListItemResource(index: number): string {
        if (index >= this._messages.length) {
            return "";
        }
        const msg = this._messages[index];
        return msg.from_me ? "ui://Emoji/chatRight" : "ui://Emoji/chatLeft";
    }

    private _addMsg(sender: string, senderIcon: string, text: string, fromMe: boolean): void {
        const sp = this._list?.getScrollPane() ?? null;
        const isScrollBottom = sp != null && sp.isBottomMost();

        const msg: ChatMessage = {
            sender,
            sender_icon: senderIcon,
            text,
            from_me: fromMe,
        };
        this._messages.push(msg);

        if (fromMe && (this._messages.length === 1 || randf() < 0.5)) {
            this._messages.push({
                sender: "FairyGUI",
                sender_icon: "r1",
                text: "Today is a good day. [:cool]",
                from_me: false,
            });
        }

        if (this._messages.length > 100) {
            this._messages = this._messages.slice(-100);
        }

        if (this._list != null) {
            this._list.setNumItems(this._messages.length);
            if (isScrollBottom && sp != null) {
                sp.scrollBottom(true);
            }
        }
    }
}

File: examples/csharp/Scripts/ChatScene.cs
using Godot;
using System.Collections.Generic;

namespace FairyGUI.Examples;

public partial class ChatScene : DemoSceneBase
{
    private class ChatMessage
    {
        public string Sender;
        public string SenderIcon;
        public string Text;
        public bool FromMe;
    }

    private GObject _list;
    private GObject _input;
    private readonly List<ChatMessage> _messages = new();
    private GObject _emojiSelectUi;

    protected override void ContinueInit()
    {
        UIPackage.AddPackage("res://Resources/UI/Emoji");

        _view = UIPackage.CreateObject("Emoji", "Main");
        _groot.AddChild(_view);

        _list = _view.GetChild("list");
        if (_list != null)
        {
            _list.SetVirtual();
            _list.SetItemRenderer(new Callable(this, MethodName.RenderListItem));
            _list.SetItemProvider(new Callable(this, MethodName.GetListItemResource));
        }

        _input = _view.GetChild("input");
        _input?.AddEventListener(FguiEvent.Submit, new Callable(this, MethodName.OnSubmit));

        var sendBtn = _view.GetChild("btnSend");
        sendBtn?.AddClickListener(new Callable(this, MethodName.OnClickSendBtn));

        var emojiBtn = _view.GetChild("btnEmoji");
        emojiBtn?.AddClickListener(new Callable(this, MethodName.OnClickEmojiBtn));

        _emojiSelectUi = UIPackage.CreateObject("Emoji", "EmojiSelectUI");
        if (_emojiSelectUi != null)
        {
            var emojiList = _emojiSelectUi.GetChild("list");
            emojiList?.AddEventListener(FguiEvent.ClickItem, new Callable(this, MethodName.OnClickEmoji));
        }

        AddMsg("Unity", "r0", "Hello!", true);
    }

    private void OnClickSendBtn()
    {
        if (_input == null)
            return;
        var msg = _input.GetText();
        if (string.IsNullOrEmpty(msg))
            return;
        AddMsg("Unity", "r0", msg, true);
        _input.SetText("");
    }

    private void OnClickEmojiBtn()
    {
        if (_emojiSelectUi == null)
            return;
        _groot.ShowPopupSimple(_emojiSelectUi);
    }

    private void OnClickEmoji()
    {
        var item = _groot.GetTouchTarget();
        if (item == null || _input == null)
            return;
        _input.SetText(_input.GetText() + "[: " + item.GetText() + "]");
    }

    private void OnSubmit()
    {
        OnClickSendBtn();
    }

    private string ParseEmoji(string text)
    {
        var regex = new RegEx();
        regex.Compile("\\[:\\s*(\\w+)\\]");
        var result = regex.Search(text);
        if (result == null)
            return text;
        var tag = result.GetString(1);
        return text.Replace(result.GetString(0), $"<img src='ui://Emoji/{tag}'/>");
    }

    private void RenderListItem(int index, GodotObject obj)
    {
        if (index >= _messages.Count)
            return;
        var msg = _messages[index];
        obj.SetIcon("ui://Emoji/" + msg.SenderIcon);

        var tf = obj.GetChild("msg");
        if (tf is GRichTextField rtf)
        {
            rtf.SetUBBEnabled(false);
            rtf.SetWidth(rtf.GetInitSize().X);
            rtf.SetText(ParseEmoji(msg.Text));
            rtf.SetWidth(rtf.GetTextSize().X);
        }

        var nameObj = obj.GetChild("name");
        if (nameObj != null && !msg.FromMe)
            nameObj.SetText(msg.Sender);
    }

    private string GetListItemResource(int index)
    {
        if (index >= _messages.Count)
            return "";
        return _messages[index].FromMe ? "ui://Emoji/chatRight" : "ui://Emoji/chatLeft";
    }

    private void AddMsg(string sender, string senderIcon, string text, bool fromMe)
    {
        var sp = _list.GetScrollPane();
        var isScrollBottom = sp != null && sp.IsBottomMost();

        _messages.Add(new ChatMessage
        {
            Sender = sender,
            SenderIcon = senderIcon,
            Text = text,
            FromMe = fromMe
        });

        if (fromMe && (_messages.Count == 1 || GD.Randf() < 0.5f))
        {
            _messages.Add(new ChatMessage
            {
                Sender = "FairyGUI",
                SenderIcon = "r1",
                Text = "Today is a good day. [:cool]",
                FromMe = false
            });
        }

        if (_messages.Count > 100)
            _messages.RemoveRange(0, _messages.Count - 100);

        if (_list != null)
        {
            _list.SetNumItems(_messages.Count);
            if (isScrollBottom && sp != null)
                sp.ScrollBottom(true);
        }
    }
}
